tests/test_spin_planewave.py

```diff
@@ -0,0 +1,293 @@
+    sim.external_fields = [
+        ExternalField(laser_func, 'Ex', a0*m_e*c**2*k0/e, lambda0),
+        ExternalField(laser_func, 'By', a0*m_e*c*k0/e, lambda0)]
+
+    # Set positions of all particles to origin
+    elec2.x = -lambda0*np.ones_like(elec2.x)
+    elec2.y = -lambda0*np.ones_like(elec2.y)
+    elec2.z = -lambda0*np.ones_like(elec2.z)
+
+    # Prepare the arrays for the time history of the pusher
+    # Nptcl = sim.ptcl[0].Ntot
+    x = np.zeros(N_step)
+    y = np.zeros(N_step)
+    z = np.zeros(N_step)
+    ux = np.zeros(N_step)
+    uy = np.zeros(N_step)
+    uz = np.zeros(N_step)
+    sx = np.zeros(N_step)
+    sy = np.zeros(N_step)
+    sz = np.zeros(N_step)
+
+    # Push the particles over N_step and record the corresponding history
+    j = 0  # which particle we select is irrelevant
+    for i in range(N_step):
+        # Record the history
+        x[i] = elec2.x[j]
+        y[i] = elec2.y[j]
+        z[i] = elec2.z[j]
+        ux[i] = elec2.ux[j]
+        uy[i] = elec2.uy[j]
+        uz[i] = elec2.uz[j]
+        sx[i] = elec2.spin_tracker.sx[j]
+        sy[i] = elec2.spin_tracker.sy[j]
+        sz[i] = elec2.spin_tracker.sz[j]
+        # Take a simulation step
+        sim.step(1, show_progress=False)
+
+    # (3) COMPUTE THE ANALYTIC SOLUTIONS
+
+    # time - use half steps here for easier comparison
+    t = sim.dt * np.arange(N_step )
+    # Setup arrays for analytic solutions, for momentum and spin
+    ux_analytic = np.zeros(N_step)
+    uz_analytic = np.zeros(N_step)
+    sx_analytic = np.zeros(N_step)
+    sz_analytic = np.zeros(N_step)
+
+    for i in range(N_step):
+        # phase
+        phi = k0*(c * t[i] - z[i])
+        # conserved quantity in plane wave propagating along z (where u_0=uz_i)
+        delta = math.sqrt(1 + uz_i ** 2) - uz_i
+        # Calculate analytic solutions
+        a0sinphi = a0 * np.sin(phi + lambda0 / 4)
+        ux_analytic[i] = a0sinphi
+        uz_analytic[i] = uz_i + 0.5 * a0sinphi ** 2 / delta
+        I = anom*a0sinphi + 2 * np.arctan(a0sinphi / (1+delta))
+        sz_analytic[i] = sz_i * np.cos(I) + sx_i * np.sin(I)
+        sx_analytic[i] = sx_i * np.cos(I) - sz_i * np.sin(I)
+
+    # Show the results
+    if show is False:
+        # Automatically check that the fields agree,
+        # to an absolute tolerance
+        atol = 0.01
+        rtol = 1e-3
+        # Interpolate to shift the analytic result by half time-step
+        ux_anal_shifted = np.interp(t[1:]-dt*0.5, t, ux_analytic)
+        assert np.allclose( ux_anal_shifted, ux[1:], atol=atol, rtol=rtol )
+        print(f'The momentum ux agrees with the theory to atol={atol},\n' +
+              'over the whole simulation box.')
+        sx_anal_shifted = np.interp(t[1:] - dt * 0.5, t, sx_analytic)
+        assert np.allclose(sx_anal_shifted, sx[1:], atol=atol, rtol=rtol)
+        print(f'The spin sx agrees with the theory to atol={atol},\n' +
+              'over the whole simulation box.')
```

Comment:
We only check one electron, so did you mean "over the duration of the simulation"?
```suggestion
        print(f'The momentum ux agrees with the theory to atol={atol},\n' +
              'over the whole duration of the simulation.')
        sx_anal_shifted = np.interp(t[1:] - dt * 0.5, t, sx_analytic)
        assert np.allclose(sx_anal_shifted, sx[1:], atol=atol, rtol=rtol)
        print(f'The spin sx agrees with the theory to atol={atol},\n' +
              'over the whole duration of the simulation.')
```